### PR TITLE
Simple ASN1 utils for DSA based on PACKET

### PR DESCRIPTION
**Pull request by mattcaswell**
This is based on the PR by @davidmakepeace in #9067. It uses all the same commits as there, but then adds some additional ones to convert the code to use the PACKET API instead.

This is WIP because I have not yet tackled WPACKET. I ran out of time today. Hopefully it is enough for now to get a feel for what it would look like. I'll work on WPACKET next week.

### CODE REVIEW

**Comment by mattcaswell** [comment]
Note: the interesting changes are in the asn1_dsa.c file in the decode* functions.

File: crypto/asn1_dsa.c [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        return 0;
+
+    return 1;
+}
+
+static int decode_dsa_sig_content(PACKET *pkt, BIGNUM *r, BIGNUM *s)
+{
+    if (!decode_der_integer(pkt, r)
+            || !decode_der_integer(pkt, s)
+            || !PACKET_remaining(pkt) == 0)
```
Comment (richsalz) [inline crypto/asn1_dsa.c]: that last clause looks strange.  perhaps a comment?

Comment (mattcaswell) [reply crypto/asn1_dsa.c]: Oh, it looked strange because it wasn't quite right. Fixed...and I actually ended up removing this whole function anyway.

**Comment by paulidale** [comment]
I'd like to see the encode side, this is where WPACKET doesn't match up with ASN.1 so nicely.

Would a fuller [ropes](https://en.wikipedia.org/wiki/Rope_(data_structure)) implementation be worthwhile?  With the variable length representation of ASN.1, it might be reasonable.

In my ideal world, ASN.1 decoding would be done using a grammar and a generated parser.

**Comment by richsalz** [comment]
I agree the encode side is the harder part, particularly for *nested* structures, as opposed to a simple sequence. Do we know what ASN1 we will have to have in the FIPS module?  I am against a fuller ropes implementation until we know it's needed. You can always allocate the max space, keep a stack, and backpatch and squeeze up for example.

**Comment by levitte** [comment]
> particularly for *nested* structures, as opposed to a simple sequence.

... although, if DSA sigs are the only use case, maybe there's not much to worry about...
(at some point, *someone* will have to decide whether we're just going to do the simplest thing for what we see right this moment, or if we're going to make something a bit more feature complete)

**Comment by richsalz** [comment]
For an example of the patch/backtrack method, see http://www.openldap.org/devel/gitweb.cgi?p=openldap.git;a=blob;f=libraries/liblber/encode.c;h=8f6d4f6ea0c62a88ed54f1226b198217cacc1624;hb=HEAD#l360

I think the Brisbane team can easily post a comprehensive list of what we need; decode RSA-PSS parameters?

**Comment by mattcaswell** [comment]
I updated this to use the WPACKET API as well. As part of that I gave WPACKET the ability to write to a NULL buffer so the number of bytes required can be calculated before later initialising the buffer.

**Comment by mattcaswell** [comment]
Update pushed addressing the comment rewording suggestion.

**Comment by mattcaswell** [comment]
Taken this out of "WIP"

File: crypto/asn1_dsa.c [hunk 2]
```diff
@@ -59,7 +59,7 @@ int encode_der_length(WPACKET *pkt, size_t cont_len)
 /*
  * Outputs the DER encoding of a positive ASN.1 INTEGER to pkt.
  *
- * If n is negative or the integer is too large then an error results.
+ * Returns an error if n is negative or too large.
```
Comment (levitte) [inline crypto/asn1_dsa.c]: "Results in... " rather than "Returns"

Comment (mattcaswell) [reply crypto/asn1_dsa.c]: Changed.

**Comment by mattcaswell** [comment]
Rebased to resolve a conflict with master.

Ping?

**Comment by paulidale** [comment]
I'm okay to approve both this and #9067.  Has a decision been made as to which is preferred?

**Comment by mattcaswell** [comment]
> Has a decision been made as to which is preferred?

Well no decision has been made as such. My own view is I strongly prefer the PACKET/WPACKET approach. Outside of that there hasn't been much discussion. My -1 on #9067 still stands. No one so far has seen fit to take that one to a vote. If an OMC member wants to take that to a vote then I'd be happy to take the debate to the OMC. Otherwise, I'd be happy to take an approval of this forward without a vote.

**Comment by richsalz** [comment]
> My -1 on #9067 

I think it's wrong to block a PR without being willing to call a vote.

**Comment by paulidale** [comment]
Merged to master after removing end of line trailing space.

**Pull request closed without merging**